compiler/qsc_data_structures/src/namespaces.rs

```diff
@@ -0,0 +1,304 @@
+    }
+
+    /// Finds the ID of a namespace given its string name. This function is generally more efficient
+    /// than [`NamespaceTreeNode::find_namespace_by_id`], as it utilizes the prefix tree structure to
+    /// find the ID in `O(n)` time, where `n` is the number of components in the namespace name.
+    pub fn get_namespace_id<'a>(
+        &self,
+        ns: impl IntoIterator<Item = &'a str>,
+    ) -> Option<NamespaceId> {
+        let mut buf: Option<NamespaceTreeCell> = None;
+        for component in ns {
+            if let Some(next_ns) = match buf {
+                None => self.get(&Rc::from(component)),
+                Some(buf) => buf.borrow().get(&Rc::from(component)),
+            } {
+                buf = Some(next_ns);
+            } else {
+                return None;
+            }
+        }
+        Some(buf.map_or_else(|| self.id, |x| x.borrow().id))
+    }
+
+    /// Inserts a new namespace into the tree, if it does not yet exist.
+    /// Returns the ID of the namespace.
+    pub fn insert_or_find_namespace<I>(
+        &mut self,
+        mut iter: Peekable<I>,
+        assigner: &mut usize,
+    ) -> Option<NamespaceId>
+    where
+        I: Iterator<Item = Rc<str>>,
+    {
+        let next_item = iter.next()?;
+        let next_node = self.children.get_mut(&next_item);
+        match (next_node, iter.peek()) {
+            (Some(next_node), Some(_)) => {
+                return next_node
+                    .borrow_mut()
+                    .insert_or_find_namespace(iter, assigner);
+            }
+            (Some(next_node), None) => {
+                return Some(next_node.borrow().id);
+            }
+            _ => {}
+        }
+        *assigner += 1;
+        let mut new_node =
+            NamespaceTreeNode::new(NamespaceId::new(*assigner), FxHashMap::default());
+        if iter.peek().is_none() {
+            let new_node_id = new_node.id;
+            self.children
+                .insert(next_item, Rc::new(RefCell::new(new_node)));
+            Some(new_node_id)
+        } else {
+            let id = new_node.insert_or_find_namespace(iter, assigner);
+            self.children
+                .insert(next_item, Rc::new(RefCell::new(new_node)));
+            id
+        }
+    }
+
+    fn find_namespace_by_id(
+        &self,
+        id: NamespaceId,
+        names_buf: &[Rc<str>],
+    ) -> Option<(Vec<Rc<str>>, NamespaceTreeCell)> {
+        // first, check if any children are the one we are looking for
+        for (name, node) in &self.children {
+            if node.borrow().id == id {
+                let mut names = names_buf.to_vec();
+                names.push(name.clone());
+                return Some((names, node.clone()));
+            }
+        }
+
+        // if it wasn't found, recurse into children
+        for (name, node) in &self.children {
+            let mut names = names_buf.to_vec();
+            names.push(name.clone());
```

Comment:
Just checking, does the `names_buf.to_vec()` create a clone of the names, such that the `names.push(name.clone())` won't pollute into the next iteration of the for loop?